state/cluster/badger/mutator.go

```diff
@@ -218,3 +216,81 @@ func (m *MutableState) Extend(block *cluster.Block) error {
 	}
 	return nil
 }
+
+// checkDupeTransactionsInUnfinalizedAncestry checks for duplicate transactions in the un-finalized
+// ancestry of the given block, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInUnfinalizedAncestry(block *cluster.Block, includedTransactions map[flow.Identifier]struct{}, finalHeight uint64) ([]flow.Identifier, error) {
+
+	var duplicateTxIDs []flow.Identifier
+	err := fork.TraverseBackward(m.headers, block.Header.ParentID, func(ancestor *flow.Header) error {
+		payload, err := m.payloads.ByBlockID(ancestor.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
+		}
+
+		for _, tx := range payload.Collection.Transactions {
+			txID := tx.ID()
+			_, duplicated := includedTransactions[txID]
+			if duplicated {
+				duplicateTxIDs = append(duplicateTxIDs, txID)
+			}
+		}
+		return nil
+	}, fork.ExcludingHeight(finalHeight))
+
+	return duplicateTxIDs, err
+}
+
+// checkDupeTransactionsInFinalizedAncestry checks for duplicate transactions in the finalized
+// ancestry, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInFinalizedAncestry(includedTransactions map[flow.Identifier]struct{}, minRefHeight, maxRefHeight uint64) ([]flow.Identifier, error) {
+	var duplicatedTxIDs []flow.Identifier
+
+	// Let E be the global transaction expiry constant, measured in blocks. For each
+	// T ∈ `includedTransactions`, we have to decide whether the transaction
+	// already appeared in _any_ finalized cluster block.
+	// Notation:
+	//   - consider a valid cluster block C and let c be its reference block height
+	//   - consider a transaction T ∈ `includedTransactions` and let t denote its
+	//     reference block height
+	//
+	// Boundary conditions:
+	// 1. C's reference block height is equal to the lowest reference block height of
+	//    all its constituent transactions. Hence, C can contain T if and only if c <= t.
+	// 2. a transaction with reference block height T is eligible for inclusion in
+	//    any collection with reference block height C where C<=T<C+E
+	//
+	// Therefore, to guarantee that we find all possible duplicates, we need to check
+	// all collections with reference block height in the range (C-E,C+E). Since we
```

Comment:
* I would suggest to update point 2 to conform to the notation.
* The argument for the overall range is slightly confusing to me.

```suggestion
	// 2. For T to be eligible for inclusion in collection C, _none_ of the transactions within C are allowed 
	// to be expired w.r.t. C's reference block. Hence, for collection C to potentially contain T, it must satisfy t < c + E.  
	//
	// Therefore, for collection C to potentially contain transaction T, it must satisfy t - E < c <= t. In other words, 
	// we only need to inspect collections with reference block height c ∈ (t-E, t]. Consequently, for a set of transactions, 
	// with `minRefHeight` (`maxRefHeight`) being the smallest (largest) reference block height, we only need to inspect collections with c ∈ (minRefHeight-E, maxRefHeight].
```